### ISSUE TITLE
Requests with bad status code generate traceback

### ISSUE THREAD

**Issue by jaimehrubiks** [synackray/vcenter-netbox-sync#41]
### Environment
* Python version:  Python 3.7.6 (also tried Python 3.7.5)
* NetBox version: v2.6.12
* vCenter version: vSphere Client version 6.7.0
* vCenter NetBox Sync version:  Tried branches: master & develop

### Steps to Reproduce
1. setup settings.py
2. python3 run.py

### Expected Behavior
No errors

### Observed Behavior
It works up until a point. It populates some information on Netbox, but it stops because of an error:

2020-01-29 22:25:10,889 [DEBUG] vc_data[slug] and nb_data[slug] values match.
2020-01-29 22:25:10,889 [DEBUG] vc_data[part_number] and nb_data[part_number] values match.
2020-01-29 22:25:10,889 [DEBUG] vc_data[tags] and nb_data[tags] key 'tags' contains list. Validating dict1 items exist in dict2.
2020-01-29 22:25:10,890 [DEBUG] vc_data[tags] and nb_data[tags] values match.
2020-01-29 22:25:10,890 [INFO] NetBox device_types object 'ProLiant DL380 Gen9' match current values. Moving on.
2020-01-29 22:25:10,890 [INFO] Finished sync of 2 vCenter hosts objects to NetBox device_types objects.
2020-01-29 22:25:10,890 [INFO] Starting sync of 10 vCenter hosts objects to NetBox devices objects.
2020-01-29 22:25:10,890 [DEBUG] Sending GET to 'http://10.102.140.212:80/api/dcim/devices/?name=esxi-p01r03u12.nix.cccis.com'
2020-01-29 22:25:10,977 [DEBUG] NetBox GET request OK; returned 200 status.
2020-01-29 22:25:10,978 [DEBUG] NetBox devices object 'esxi-p01r03u12.nix.cccis.com' already exists. Comparing values.
2020-01-29 22:25:10,978 [DEBUG] vc_data[name] and nb_data[name] values match.
2020-01-29 22:25:10,978 [DEBUG] vc_data[device_type] and nb_data[device_type] contain dictionary. Evaluating.
2020-01-29 22:25:10,978 [DEBUG] vc_data[device_type][model] and nb_data[device_type][model] values match.
2020-01-29 22:25:10,979 [DEBUG] vc_data[device_type] and nb_data[device_type] values match.
2020-01-29 22:25:10,979 [DEBUG] vc_data[device_role] and nb_data[device_role] contain dictionary. Evaluating.
2020-01-29 22:25:10,979 [DEBUG] vc_data[device_role][name] and nb_data[device_role][name] values match.
2020-01-29 22:25:10,979 [DEBUG] vc_data[device_role] and nb_data[device_role] values match.
2020-01-29 22:25:10,979 [DEBUG] vc_data[platform] and nb_data[platform] contain dictionary. Evaluating.
2020-01-29 22:25:10,980 [DEBUG] vc_data[platform][name] and nb_data[platform][name] values match.
2020-01-29 22:25:10,980 [DEBUG] vc_data[platform] and nb_data[platform] values match.
2020-01-29 22:25:10,980 [DEBUG] vc_data[site] and nb_data[site] contain dictionary. Evaluating.
2020-01-29 22:25:10,980 [DEBUG] vc_data[site][name] and nb_data[site][name] values match.
2020-01-29 22:25:10,980 [DEBUG] vc_data[site] and nb_data[site] values match.
2020-01-29 22:25:10,981 [DEBUG] vc_data[serial] and nb_data[serial] values match.
2020-01-29 22:25:10,981 [DEBUG] vc_data[asset_tag] and nb_data[asset_tag] values match.
2020-01-29 22:25:10,981 [DEBUG] vc_data[cluster] and nb_data[cluster] contain dictionary. Evaluating.
2020-01-29 22:25:10,981 [DEBUG] vc_data[cluster][name] and nb_data[cluster][name] values match.
2020-01-29 22:25:10,981 [DEBUG] vc_data[cluster] and nb_data[cluster] values match.
2020-01-29 22:25:10,982 [DEBUG] vc_data[status] and nb_data[status] values match.
2020-01-29 22:25:10,982 [DEBUG] vc_data[tags] and nb_data[tags] key 'tags' contains list. Validating dict1 items exist in dict2.
2020-01-29 22:25:10,982 [DEBUG] vc_data[tags] and nb_data[tags] values match.
2020-01-29 22:25:10,983 [INFO] NetBox devices object 'esxi-p01r03u12.nix.cccis.com' match current values. Moving on.
2020-01-29 22:25:10,983 [DEBUG] Sending GET to 'http://10.102.140.212:80/api/dcim/devices/?name=esxi-p01r03u10.nix.cccis.com'
2020-01-29 22:25:11,049 [DEBUG] NetBox GET request OK; returned 200 status.
2020-01-29 22:25:11,049 [DEBUG] NetBox devices object 'esxi-p01r03u10.nix.cccis.com' already exists. Comparing values.
2020-01-29 22:25:11,049 [DEBUG] vc_data[name] and nb_data[name] values match.
2020-01-29 22:25:11,049 [DEBUG] vc_data[device_type] and nb_data[device_type] contain dictionary. Evaluating.
2020-01-29 22:25:11,049 [DEBUG] vc_data[device_type][model] and nb_data[device_type][model] values match.
2020-01-29 22:25:11,050 [DEBUG] vc_data[device_type] and nb_data[device_type] values match.
2020-01-29 22:25:11,050 [DEBUG] vc_data[device_role] and nb_data[device_role] contain dictionary. Evaluating.
2020-01-29 22:25:11,050 [DEBUG] vc_data[device_role][name] and nb_data[device_role][name] values match.
2020-01-29 22:25:11,050 [DEBUG] vc_data[device_role] and nb_data[device_role] values match.
2020-01-29 22:25:11,051 [DEBUG] vc_data[platform] and nb_data[platform] contain dictionary. Evaluating.
2020-01-29 22:25:11,051 [DEBUG] vc_data[platform][name] and nb_data[platform][name] values match.
2020-01-29 22:25:11,051 [DEBUG] vc_data[platform] and nb_data[platform] values match.
2020-01-29 22:25:11,051 [DEBUG] vc_data[site] and nb_data[site] contain dictionary. Evaluating.
2020-01-29 22:25:11,051 [DEBUG] vc_data[site][name] and nb_data[site][name] values match.
2020-01-29 22:25:11,052 [DEBUG] vc_data[site] and nb_data[site] values match.
2020-01-29 22:25:11,052 [DEBUG] vc_data[serial] and nb_data[serial] values match.
2020-01-29 22:25:11,052 [DEBUG] vc_data[asset_tag] and nb_data[asset_tag] values match.
2020-01-29 22:25:11,052 [DEBUG] vc_data[cluster] and nb_data[cluster] contain dictionary. Evaluating.
2020-01-29 22:25:11,052 [DEBUG] vc_data[cluster][name] and nb_data[cluster][name] values match.
2020-01-29 22:25:11,052 [DEBUG] vc_data[cluster] and nb_data[cluster] values match.
2020-01-29 22:25:11,052 [DEBUG] vc_data[status] and nb_data[status] values match.
2020-01-29 22:25:11,052 [DEBUG] vc_data[tags] and nb_data[tags] key 'tags' contains list. Validating dict1 items exist in dict2.
2020-01-29 22:25:11,053 [DEBUG] vc_data[tags] and nb_data[tags] values match.
2020-01-29 22:25:11,053 [INFO] NetBox devices object 'esxi-p01r03u10.nix.cccis.com' match current values. Moving on.
2020-01-29 22:25:11,053 [DEBUG] Sending GET to 'http://10.102.140.212:80/api/dcim/devices/?name=esxi-p01r02u12.nix.cccis.com'
2020-01-29 22:25:11,090 [DEBUG] NetBox GET request OK; returned 200 status.
2020-01-29 22:25:11,090 [INFO] Netbox devices 'esxi-p01r02u12.nix.cccis.com' object not found. Requesting creation.
2020-01-29 22:25:11,091 [DEBUG] Sending POST to 'http://10.102.140.212:80/api/dcim/devices/'
Traceback (most recent call last):
  File "run.py", line 1240, in <module>
    main()
  File "run.py", line 48, in main
    nb.sync_objects(vc_obj_type="hosts")
  File "run.py", line 938, in sync_objects
    self.obj_exists(nb_obj_type=nb_obj_type, vc_data=obj)
  File "run.py", line 886, in obj_exists
    req_type="post", nb_obj_type=nb_obj_type, data=vc_data
  File "run.py", line 792, in request
    req_type.upper(), req.status_code, data, req.json()
  File "/usr/local/lib/python3.7/site-packages/requests/models.py", line 897, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/local/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python3.7/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 2 column 1 (char 1)

**Comment by synackray** [comment]
Hello @jaimehrubiks! It's a pleasure to meet you. Seeing as this is failing on the creation of a new object, I'll need to see the object details to try and modify the response collection to see if we can sort out why it's not crashing. 

Give me a few moments and I'll push a branch that we can do further testing against. Thanks!

**Comment by synackray** [comment]
Would you kindly pull the latest commit from `41-request-json-failure` branch and re-run and then provide the debug data? Thank you so much!

**Comment by jaimehrubiks** [comment]
Great!, I'll try soon, will let you know.

**Comment by jaimehrubiks** [comment]
@synackray 
https://pastebin.com/w059Cra5

**Comment by synackray** [comment]
AHA!

```
duplicate key value violates unique constraint &quot;dcim_device_asset_tag_key&quot;
DETAIL:  Key (asset_tag)=() already exists.
```

The asset tag field is an empty string and is not being properly caught in the banned list. Let me work on an update and get back to you with an update to test.

**Comment by synackray** [comment]
@jaimehrubiks Do you mind pulling the latest commit and giving it a shot? I tested locally by overwriting my asset tags with empty strings and it seems to function, but I'd like your feedback. Thanks!

**Comment by jaimehrubiks** [comment]
@synackray 
Thanks, it works great now. And thanks for this amazing tool
So what is asset_tag supposed to be used for? Why don't I have mine populated from vCenter

**Comment by synackray** [comment]
Some companies place asset tags on all of their hardware in order to track their inventory and well.. assets. It is typically a unique serial number or identifier placed onto the hardware. If your company does this, and then populates the data into the hardware BIOS or host configuration then it would be picked up and placed in the NetBox host model asset tag field.

I'm glad this works! I'll start moving this towards `develop` and into `master`.

**Comment by jaimehrubiks** [comment]
I'm on the latest develop branch
commit b157290c3c27938e7a295327cf70745bde174f05 (HEAD -> develop, origin/develop)

This seems to be the same error, isn't it?


2020-02-10 11:26:24,822 [DEBUG] Received HTTP Status 200.
2020-02-10 11:26:24,822 [DEBUG] NetBox GET request OK; returned 200 status.
2020-02-10 11:26:24,822 [INFO] Netbox devices 'esxi-p01r01u10.nix.cccis.com' object not found. Requesting creation.
2020-02-10 11:26:24,822 [DEBUG] Sending POST to 'http://10.102.140.212:80/api/dcim/devices/' with data '{'name': 'esxi-p01r01u10.nix.cccis.com', 'device_role': {'name': 'Server'}, 'device_type': {'model': 'ProLiant DL380 Gen9'}, 'platform': {'name': 'VMware ESXi'}, 'site': {'name': 'vCenter'}, 'serial': 'blurred', 'asset_tag': '                                ', 'cluster': {'name': 'Main-Cluster'}, 'status': 1, 'tags': ['Synced', 'vCenter', '10.102.138.31']}'.
2020-02-10 11:26:24,983 [DEBUG] Received HTTP Status 500.
2020-02-10 11:26:24,983 [CRITICAL] Well this in unexpected. Please report this. POST request received 500 status with body '{'name': 'esxi-p01r01u10.nix.cccis.com', 'device_role': {'name': 'Server'}, 'device_type': {'model': 'ProLiant DL380 Gen9'}, 'platform': {'name': 'VMware ESXi'}, 'site': {'name': 'vCenter'}, 'serial': 'blurred', 'asset_tag': '                                ', 'cluster': {'name': 'Main-Cluster'}, 'status': 1, 'tags': ['Synced', 'vCenter', '10.102.138.31']}' and response '
<!DOCTYPE html>
<html lang="en">

<head>
    <title>Server Error</title>
    <link rel="stylesheet" href="/static/bootstrap-3.4.1-dist/css/bootstrap.min.css">
    <link rel="stylesheet" href="/static/font-awesome-4.7.0/css/font-awesome.min.css">
    <meta charset="UTF-8">
</head>

<body>
    <div class="container-fluid">
        <div class="row">
            <div class="col-md-6 col-md-offset-3">
                <div class="panel panel-danger" style="margin-top: 200px">
                    <div class="panel-heading">
                        <strong>
                            <i class="fa fa-warning"></i>
                            Server Error
                        </strong>
                    </div>
                    <div class="panel-body">

                            <p>
                                There was a problem with your request. Please contact an administrator.
                            </p>

                        <hr />
                        <p>
                            The complete exception is provided below:
                        </p>
<pre><strong>&lt;class &#39;django.db.utils.IntegrityError&#39;&gt;</strong><br />
duplicate key value violates unique constraint &quot;dcim_device_asset_tag_key&quot;
DETAIL:  Key (asset_tag)=() already exists.
</pre>
                        <p>
                            If further assistance is required, please post to the <a href="https://groups.google.com/forum/#!forum/netbox-discuss">NetBox mailing list</a>.
                        </p>
                        <div class="text-right">
                            <a href="/" class="btn btn-primary">Home Page</a>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</body>

</html>
'.

**Comment by synackray** [comment]
Oh no! How did this AT get through? Let me verify real quick.. In theory `'asset_tag': ' '` should not be allowed.

**Comment by synackray** [comment]
I continue to struggle to understand how this one slips through. I feel like I'm playing a cat and mouse game. :)

Commit b157290c3c27938e7a295327cf70745bde174f05 contains the following banned asset tags
`banned_tags = ["Default string", "Unknown", " ", ""]`

Upon collection of the asset tag the it's evaluated.
```
if asset_tag in banned_tags:
    log.debug("Banned asset tag string. Nulling.")
    asset_tag = None
```
So I try this manually, pasting what output you've provided.

```
$ python3 
Python 3.6.9 (default, Nov  7 2019, 10:44:02) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> asset_tag = ' '
>>> banned_tags = ["Default string", "Unknown", " ", ""]
>>> banned_tags = [t.lower() for t in banned_tags]
>>> asset_tag in banned_tags
True
>>> 
```
I get the result expected. The asset tag should be nulled. I want to be careful of being overly loose with banned strings as we have no idea what formats may exist. Let me write up a few debug changes just to make sure I'm not losing my mind.

**Comment by synackray** [comment]
@jaimehrubiks Would you mind providing the vCenter host collection portion of the logs? For the host in question did you get a message about the tag being in the banned list and being nulled? I've also committed f0799b139222c49ce29f542735370879505f5eb5 if you'd like to run it and provide the debugs. It has additional steps for logging.

P.S. I can recreate the issue by overwriting the final result of the ban, but in theory that should not be occurring.

```
2020-02-10 19:08:35,916 [CRITICAL] Well this in unexpected. Please report this. POST request received 500 status with body '{'name': 'vcenter1.lab1.local', 'device_role': {'name': 'Server'}, 'device_type': {'model': 'XXXX'}, 'platform': {'name': 'VMware ESXi'}, 'site': {'name': 'vCenter'}, 'serial': 'XXXX', 'asset_tag': ' ', 'cluster': {'name': 'c2-b62-dfw3-Exchange'}, 'status': 'active', 'tags': ['Synced', 'vCenter', '192.168.1.11']}'

<pre><strong>&lt;class &#39;django.db.utils.IntegrityError&#39;&gt;</strong><br />
duplicate key value violates unique constraint &quot;dcim_device_asset_tag_key&quot;
DETAIL:  Key (asset_tag)=() already exists.
</pre>
```

**Comment by jaimehrubiks** [comment]
I know it looks super weird haha but some of my tags are like multiple spaces. I don't really now why.

2020-02-10 14:15:34,669 [DEBUG] Collecting info to create NetBox manufacturers object.
2020-02-10 14:15:34,669 [DEBUG] Collecting info to create NetBox device_types object.
2020-02-10 14:15:34,669 [DEBUG] Collecting info to create NetBox devices object.
2020-02-10 14:15:34,702 [DEBUG] Received asset tag '' from vCenter.
2020-02-10 14:15:34,702 [DEBUG] Banned asset tag string. Nulling.
2020-02-10 14:15:34,703 [DEBUG] Final decided asset tag: None
2020-02-10 14:15:34,742 [DEBUG] Collecting info to create NetBox interfaces object.
2020-02-10 14:15:35,297 [DEBUG] Collecting info for physical interface 'vmnic0'.
2020-02-10 14:15:35,297 [DEBUG] Collecting info for physical interface 'vmnic1'.
2020-02-10 14:15:35,297 [DEBUG] Collecting info for physical interface 'vmnic2'.
2020-02-10 14:15:35,297 [DEBUG] Collecting info for physical interface 'vmnic3'.
2020-02-10 14:15:35,297 [DEBUG] Collecting info for physical interface 'vmnic4'.
2020-02-10 14:15:35,297 [DEBUG] Collecting info for physical interface 'vmnic5'.
2020-02-10 14:15:35,297 [DEBUG] Collecting info for physical interface 'vusb0'.
2020-02-10 14:15:35,828 [DEBUG] Collecting info for virtual interface 'vmk0'.
2020-02-10 14:15:35,828 [DEBUG] Collecting info for IP Address '10.102.137.144'.
2020-02-10 14:15:35,828 [DEBUG] Collecting info for virtual interface 'vmk1'.
2020-02-10 14:15:35,828 [DEBUG] Collecting info for IP Address '10.102.160.144'.
2020-02-10 14:15:35,828 [DEBUG] Collecting info for virtual interface 'vmk2'.
2020-02-10 14:15:35,829 [DEBUG] Collecting info for IP Address '10.102.161.144'.
2020-02-10 14:15:35,829 [DEBUG] Collecting info for virtual interface 'vmk3'.
2020-02-10 14:15:35,829 [DEBUG] Collecting info for IP Address '10.102.162.144'.
2020-02-10 14:15:35,829 [DEBUG] Collecting info for virtual interface 'vmk4'.
2020-02-10 14:15:35,829 [DEBUG] Collecting info for IP Address '10.102.163.144'.
2020-02-10 14:15:35,829 [DEBUG] Collecting info for virtual interface 'vmk5'.
2020-02-10 14:15:35,830 [DEBUG] Collecting info for IP Address '10.102.134.144'.
2020-02-10 14:15:35,832 [INFO] Collecting info about vCenter hosts 'esxi-p01r03u10.nix.cccis.com' object.
2020-02-10 14:15:35,911 [DEBUG] Collecting info to create NetBox manufacturers object.
2020-02-10 14:15:35,911 [DEBUG] Collecting info to create NetBox device_types object.
2020-02-10 14:15:35,911 [DEBUG] Collecting info to create NetBox devices object.
2020-02-10 14:15:35,950 [DEBUG] Received asset tag '                                ' from vCenter.
2020-02-10 14:15:35,950 [DEBUG] Final decided asset tag:
2020-02-10 14:15:35,997 [DEBUG] Collecting info to create NetBox interfaces object.
2020-02-10 14:15:36,409 [DEBUG] Collecting info for physical interface 'vmnic0'.
2020-02-10 14:15:36,409 [DEBUG] Collecting info for physical interface 'vmnic1'.
2020-02-10 14:15:36,409 [DEBUG] Collecting info for physical interface 'vmnic2'.
2020-02-10 14:15:36,409 [DEBUG] Collecting info for physical interface 'vmnic3'.
2020-02-10 14:15:36,409 [DEBUG] Collecting info for physical interface 'vmnic4'.

**Comment by synackray** [comment]
Ahah now that's the kind of weird string I'd expect. I'm going to add in an option to skip the collection of asset tags. I'll also work on improving the logic further. Perhaps a check for multiple spaces is reasonable without limiting other users.

![image](https://user-images.githubusercontent.com/24757919/74187573-1f53cd80-4c4d-11ea-88b1-7ff2fea0d15f.png)

Thanks for all of your logs and assistance!

**Comment by synackray** [comment]
Alrighty @jaimehrubiks I've put in a more thorough is_banned_asset_tag function and I've also enabled the option to disable asset tag collections in `settings.py`. 

I'll run this through all branch tests and then promote to `develop`. Please let me know if you run into any issues!

**Comment by jaimehrubiks** [comment]
Works for me now! Thanks. Also, the option to completely disable it will be of help in case someone encounters a similar problem in the future (like having the same string in all their devices).

**Comment by synackray** [comment]
Woohoo! Thanks again for the testing.

**Comment by synackray** [comment]
Merged into `develop` on #64.
